runtime/jvmti/jvmtiHook.c

```diff
@@ -1579,18 +1579,17 @@ jvmtiHookGetEnv(J9HookInterface** hook, UDATA eventNum, void* eventData, void* u
 
 	Trc_JVMTI_jvmtiHookGetEnv_Entry();
 
-	if (data->rc == JNI_EVERSION) {
+	if (JNI_EVERSION == data->rc) {
 		jint version = data->version & ~JVMTI_VERSION_MASK_MICRO;
 
 		if ((JVMTI_VERSION_1_0 == version)
 			|| (JVMTI_VERSION_1_1 == version)
 			|| (JVMTI_VERSION_1_2 == version)
 #if JAVA_SPEC_VERSION > 8
-			|| (JVMTI_VERSION_9 == version)
-			|| (JVMTI_VERSION_11 == version)
-#if JAVA_SPEC_VERSION >= 15
-			|| (JVMTI_VERSION_15 == version)
-#endif /* JAVA_SPEC_VERSION >= 15 */
+			/* JVMTI 9+ with mino/micro versions are not accepted, this could change if such support is added explicitly */
+			|| (J9_ARE_ALL_BITS_SET(version, JVMTI_VERSION_INTERFACE_JVMTI)
```

Comment:
This is more complicated than the snippet I suggested (and incorrect besides). ALL_BITS simply cannot be used in this manner.

Reply:
Is it required to verify the `JVMTI_VERSION_INTERFACE_JVMTI` bits?
How about a version like `0x10010000`?

Reply:
This needs to extract the requested JVMTI version and validate that it is understood and supported by the running VM. Just checking the bits will not yield the right answer in all cases.

Reply:
`J9_ARE_ALL_BITS_SET(version, JVMTI_VERSION_INTERFACE_JVMTI)` verifies JVMTI interface bits, i.e., `0x30000000`.
`J9_ARE_NO_BITS_SET(version, JVMTI_VERSION_MASK_MINOR|JVMTI_VERSION_MASK_MICRO)` verifies no minor/micro bits (currently no such support for JVMTI 9+)
`(((version & JVMTI_VERSION_MASK_MAJOR) >> JVMTI_VERSION_SHIFT_MAJOR) <= JAVA_SPEC_VERSION))` verified that the major version is same or less than current version.
Together, this verifies the incoming version is supported by running VM.

Reply:
That accepts values like `0x30010300` or `0x300c0000` which are not valid (no special versions exist for 3 nor 12, for example).

Reply:
`0x30010300` is going to detected by `J9_ARE_NO_BITS_SET(version, JVMTI_VERSION_MASK_MINOR|JVMTI_VERSION_MASK_MICRO)`.
`0x300c0000` is technically correct according to current version scheme, I think it won't hurt to accept it.
To be strictly correct, it can be hardcoded to exclude `0x300c0000`.

Reply:
In my example, I meant to write `0x30030000` instead of `0x30010300`. I don't think we should accept code just because they fit a new pattern.

Your proposal also rejects these (with non-zero minor version numbers):
```
#define JVMTI_VERSION_1_1 0x30010100
#define JVMTI_VERSION_1_2 0x30010200
```

Reply:
Please use the code I slacked you.

Reply:
Actually, the simple version I sent you is wrong. Rather than using ALL_BITS, mask the version with the interface type mask and compare it to the JVMTI interface value. Along with your NO_BITS code (which needs spaces around the |) this check should allow only versions which are exactly a major version equal to the JDK version.

Reply:
> this check should allow only versions which are exactly a major version equal to the JDK version.

I think the earlier versions such as 11 & 13 need to be accepted by current 16 VM.

Reply:
After some offline discussion with GAC, the code in question now looks like
```
			jint majorVersion = ((version & JVMTI_VERSION_MASK_MAJOR) >> JVMTI_VERSION_SHIFT_MAJOR);
			isValidVersion = ((majorVersion > 8)
				&& (majorVersion <= JAVA_SPEC_VERSION)
				&& (JVMTI_VERSION_INTERFACE_JVMTI == (version & ~JVMTI_VERSION_MASK_MAJOR)));
```